### PR TITLE
Sudoku classes

### PR DESCRIPTION
**Pull request by anand2312**
Classes+Methods needed for the puzzle generation/answer checking logic completed.

### CODE REVIEW

File: sudoku-django-project/mainapp/classes.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+    def __init__(self):  # noqa: ANN204
+        self.counter = 1
+        self.top_boxes = [_Box() for _ in range(3)]
+        self.mid_boxes = [_Box() for _ in range(3)]
+        self.bottom_boxes = [_Box() for _ in range(3)]
+
+        self._set_box_locations()
+
+        self.original = [[1, 2, 3, 4, 5, 6, 7, 8, 9],
```
Comment (shriram-sg) [inline sudoku-django-project/mainapp/classes.py:17]: Can be made smaller with list comprehension

Comment (shriram-sg) [reply sudoku-django-project/mainapp/classes.py:17]: Maybe make the original a separate function, so the object's attribute only contains the actual puzzle and not the thing you start off with

Comment (anand2312) [reply sudoku-django-project/mainapp/classes.py:17]: Yeah having an attribute for the original seems unnecessary. How would a list comp work though? The first inner list is simple enough, but the rest are shuffled :thinking:

File: sudoku-django-project/mainapp/classes.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+                         [7, 8, 9, 1, 2, 3, 4, 5, 6],
+                         [2, 3, 1, 5, 6, 4, 8, 9, 7],
+                         [5, 6, 4, 8, 9, 7, 2, 3, 1],
+                         [8, 9, 7, 2, 3, 1, 5, 6, 4],
+                         [3, 1, 2, 6, 4, 5, 9, 7, 8],
+                         [6, 4, 5, 9, 7, 8, 3, 1, 2],
+                         [9, 7, 8, 3, 1, 2, 6, 4, 5]]
+
+        self.generator()
+        self.full_board = self.get_all_row_values()
```
Comment (shriram-sg) [inline sudoku-django-project/mainapp/classes.py:28]: No need to make a separate get_all_row_values() function if you use list comprehension with get_row_values() method

File: sudoku-django-project/mainapp/classes.py (hunk 1 above)
Comment (anand2312) [reply sudoku-django-project/mainapp/classes.py:17]: I'm reluctant to use list comp just for one of the inner elements and then hardcode the rest, feel like it'll look...ugly.

Comment (shriram-sg) [reply sudoku-django-project/mainapp/classes.py:17]: all the inner lists are 1-9 right, so you could do
`[[i for i in range(1,10)] for i in range(1,10)]`

Comment (risb21) [reply sudoku-django-project/mainapp/classes.py:17]: this is supposed to be a working board, i.e., should be able to be used as a puzzle itself, hence it is hard coded.

Comment (anand2312) [reply sudoku-django-project/mainapp/classes.py:17]: The order in the rest is different, but the shuffle isn't random, so we can't `random.shuffle` it (I think, I'm not very sure how this works)

Comment (shriram-sg) [reply sudoku-django-project/mainapp/classes.py:17]: So the actual puzzle after generation is stored in the original itself?

File: sudoku-django-project/mainapp/classes.py (hunk 2 above)
Comment (risb21) [reply sudoku-django-project/mainapp/classes.py:28]: alright will correct in the upcoming commit 👍

File: sudoku-django-project/mainapp/classes.py (hunk 1 above)
Comment (anand2312) [reply sudoku-django-project/mainapp/classes.py:17]: Unsure, but if so that's now a misleading variable name

Comment (shriram-sg) [reply sudoku-django-project/mainapp/classes.py:17]: Also what does the fullboard attribute hold then?

Comment (risb21) [reply sudoku-django-project/mainapp/classes.py:17]: The way the generator works is that the class object holds the shuffled version (in a way that creates a board) of the hard coded board, and it was used for shorter computation times (compared to back tracking for generation). The original board wont change value, but the instance of the class (self) itself will store the shuffled board.

Comment (shriram-sg) [reply sudoku-django-project/mainapp/classes.py:17]: so what do the original board and fullboard attributes hold exactly?

File: sudoku-django-project/mainapp/classes.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        self.puzzle_maker()
+
+    # Methods run at initialization.
+    def _set_box_locations(self) -> None:  # Sets each of the box's position attribute.
```
Comment (shriram-sg) [inline sudoku-django-project/mainapp/classes.py:32]: Can remove, I don't think it breaks anything and is unnecessary

Comment (shriram-sg) [reply sudoku-django-project/mainapp/classes.py:32]: Also put all the methods run at initialization here, I think that makes it more organized

File: sudoku-django-project/mainapp/classes.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+        for element in self:  # returns True or False.
+            if element.get_value() == val:
+                return True
+        return False
+
+
+class _Element:
+    """ A class that acts as the element(number) in one of the boxes of sudoku. """
+
+    def __init__(self, value: typing.Optional[int]=None):   # noqa: ANN204
```
Comment (shriram-sg) [inline sudoku-django-project/mainapp/classes.py:266]: Can remove the box and row attributes if they don't break anything in your code

File: sudoku-django-project/mainapp/classes.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+
+
+class _Box:
+    """ A class that acts like one of the nine 3x3 boxes in sudoku. """
+
+    def __init__(self):  # noqa: ANN204
+        self.top_row = [_Element() for _ in range(3)]
+        self.mid_row = [_Element() for _ in range(3)]
+        self.bottom_row = [_Element() for _ in range(3)]
+        self.box_pos = None  # NOT USED
```
Comment (shriram-sg) [inline sudoku-django-project/mainapp/classes.py:215]: Can remove the box pos, and the two methods that are being run here as well as their definitions, if they don't break anything in your code

File: sudoku-django-project/mainapp/classes.py (hunk 3 above)
Comment (risb21) [reply sudoku-django-project/mainapp/classes.py:32]: alright 👌

File: sudoku-django-project/mainapp/classes.py (hunk 1 above)
Comment (risb21) [reply sudoku-django-project/mainapp/classes.py:17]: @Windsmith Original stores the original board that is used to generate a separate board, full_board stores the newly generated board with all the solutions, and I guess puzzle_board is not really necessary, so I can remove that variable

Comment (shriram-sg) [reply sudoku-django-project/mainapp/classes.py:17]: puzzle_board variable?
Also I think its better if you turn the original board into a separate function that just returns that board. I doubt its good to leave multiple board configurations as attributes for a class supposed to represent a single board.
You could make it a static method.

**Review by shriram-sg — APPROVED** [review]

**Pull request merged**